### PR TITLE
[ROCm] add MIGraphX ci pipeline

### PR DESCRIPTION
**Pull request by PeixuanZuo**
**Description**: Describe your changes.
Add migraphx ci pipeline, test build and unit tests.
This PR is based on #11492 

Pipeline : https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=765711&view=results

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/linux-migraphx-ci-pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  # gid of video and render group on gcr-openpai-35 and -36
+  variables:
+    - name: video
+      value: 44
+    - name: render
+      value: 109
+
+  # generated from tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile
+  container:
+    image: onnxruntimecibuildenvironment.azurecr.io/migraphx-ci-pipeline-env
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/linux-migraphx-ci-pipeline.yml:18]: What's this? Who maintains this image?

File: tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+WORKDIR /stage
+
+# from rocm/pytorch's image, work around ucx's dlopen replacement conflicting with shared provider
+RUN cd /opt/mpi_install/ucx/build &&\
+      make clean &&\
+      ../contrib/configure-release --prefix=/opt/ucx --without-rocm &&\
+      make -j $(nproc) &&\
+      make install
+
+# CMake
```
Comment (snnn) [inline tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile:21]: Why do you need to install cmake twice?

File: tools/ci_build/github/azure-pipelines/linux-migraphx-ci-pipeline.yml (hunk 1 above)
Comment (PeixuanZuo) [reply tools/ci_build/github/azure-pipelines/linux-migraphx-ci-pipeline.yml:18]: It's a fixed docker image for MIGraphX CI, it's like the "rocm-ci-pipeline-env" in https://github.com/microsoft/onnxruntime/blob/e3a2d5cca8bcefe064f83d57e46ea51ddb2b16e8/tools/ci_build/github/azure-pipelines/orttraining-pai-ci-pipeline.yml#L20

File: tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile (hunk 2 above)
Comment (PeixuanZuo) [reply tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile:21]: cleaned it.

File: tools/ci_build/github/azure-pipelines/linux-migraphx-ci-pipeline.yml (hunk 1 above)
Comment (snnn) [reply tools/ci_build/github/azure-pipelines/linux-migraphx-ci-pipeline.yml:18]: Could you please create a different ACR for this?

Comment (PeixuanZuo) [reply tools/ci_build/github/azure-pipelines/linux-migraphx-ci-pipeline.yml:18]: I don't know exactly what this ACR was created for, but now only AMD CI is using it.

**Comment by PeixuanZuo** [comment]
@weixingzhang Could you please review this PR?

**Comment by snnn** [comment]
@askhade I assume your team is the owner of onnxruntimecibuildenvironment.azurecr.io and will take charge of maintaining the images, is right?

**Comment by ytaous** [comment]
@weixingzhang

**Comment by snnn** [comment]
My question wasn't answered.

**Comment by PeixuanZuo** [comment]
> @askhade I assume your team is the owner of onnxruntimecibuildenvironment.azurecr.io and will take charge of maintaining the images, is it right?

@baijumeswani @ytaous Could you please make sure it?

File: tools/ci_build/github/azure-pipelines/linux-migraphx-ci-pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
+trigger: none
+
+name: 'linux_ci_$(Date:yyyyMMdd)_$(Rev:r)'
+jobs:
+- job: AMDMIGraphX_CI
+  pool: 'AMD-GPU'
```
Comment (snnn) [inline tools/ci_build/github/azure-pipelines/linux-migraphx-ci-pipeline.yml:6]: Unless you can have new machines for this pool, you should let the existing users of this pool know that half of their resource will be allocated to this new pipeline.

Comment (PeixuanZuo) [reply tools/ci_build/github/azure-pipelines/linux-migraphx-ci-pipeline.yml:6]: This pipeline takes nearly half an hour, less than ROCm CI pipeline(1h 20 min). We have 8 available agents in `AMD  pool` and  only turn on 4 agents for ROCm CI pipeline now. We can turn on more agent if Migraphx pipeline causes a shortage of resources.

File: tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,38 @@
+FROM rocm/pytorch:rocm5.2.3_ubuntu20.04_py3.7_pytorch_1.12.1
```
Comment (snnn) [inline tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile:1]: How is this file used?

Comment (PeixuanZuo) [reply tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile:1]: We manually use it to generate a new docker image with latest ROCm version and push the new docker image to nnxruntimecibuildenvironment.azurecr.io. Migraphx CI pipeline uses the builded docker image directly.

File: tools/ci_build/github/pai/migraphx_test_launcher.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+echo "Warning: The following tests are EXCLUDED on MIGraphX agent:"
+gtest_filter="-"
+while read line; do
+  gtest_filter="$gtest_filter:$line"
+  echo "$line"
+done <$script_dir/migraphx-excluded-tests.txt
+echo ""
+
+echo "Running ./onnxruntime_test_all .."
+$build_dir/onnxruntime_test_all --gtest_filter=$gtest_filter
```
Comment (ytaous) [inline tools/ci_build/github/pai/migraphx_test_launcher.sh:15]: >onnxruntime_test_all  [](http://example.com/codeflow?start=11&length=21)

Does it run InferenceSessionTests.* under 
./onnxruntime/test/framework/inference_session_test.cc
? If not, perhaps in future we can consider to run those instead. If so, perhaps we can include ifdef USE_MIGRAPHX in the beginning. Right now, there's USE_ROCM there.

**Review by ytaous — DISMISSED** [review]

File: tools/ci_build/github/pai/migraphx_test_launcher.sh (hunk 5 above)
Comment (PeixuanZuo) [reply tools/ci_build/github/pai/migraphx_test_launcher.sh:15]: It's run part of InferenceSessionTests.* tests, but not included the GPU part. We can add them if we need.

File: tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile (hunk 4 above)
Comment (snnn) [reply tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile:1]: Please have a doc:
1. How to generate the image onnxruntimecibuildenvironment.azurecr.io/migraphx-ci-pipeline-env:rocm5.2.3
2. Who will do the job on a regular basis. For example, once a week.

Comment (snnn) [reply tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile:1]: The purpose of having build pipelines is to **automate** build process. Why half of the work of this pipeline needs be done manually?

Comment (PeixuanZuo) [reply tools/ci_build/github/pai/migraphx-ci-pipeline-env.Dockerfile:1]: Change the usage of Dockerfile. We build the dockerfile as a step of pipeline now, and use the same register container to store docker image as other pipeline does.

**Review by snnn — COMMENTED** [review]
LGTM. 

Thanks

**Review by snnn — APPROVED** [review]

**Review by ytaous — APPROVED** [review]

**Pull request merged**